File: backend/fastapi/app/models/events.py
"""
事件相关数据模型
"""

from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class EventData(BaseModel):
    """单个事件数据模型"""
    user_id: str
    session_id: str
    event_type: str
    page_url: str
    element: Optional[str] = None
    element_text: Optional[str] = None
    timestamp: int
    extra_data: Optional[Dict] = None


class BatchEventRequest(BaseModel):
    """批量事件请求模型"""
    events: List[EventData]


class EventResponse(BaseModel):
    """事件响应模型"""
    status: str
    message: str
    processed_count: int
    golang_response: Optional[Dict] = None 

File: backend/hono/src/routes/events.ts
/**
 * 事件相关API路由
 */

import { Hono } from 'hono';
import { z } from 'zod';
import { zValidator } from '@hono/zod-validator';
import { golangService } from '../services/golang-service';
import { analyticsService } from '../services/analytics-service';
import { EventResponse } from '../types';

const events = new Hono();

// 验证器
const eventDataSchema = z.object({
  user_id: z.string(),
  session_id: z.string(),
  event_type: z.string(),
  page_url: z.string(),
  element: z.string().optional(),
  element_text: z.string().optional(),
  timestamp: z.number(),
  extra_data: z.record(z.any()).optional(),
});

const batchEventSchema = z.object({
  events: z.array(eventDataSchema),
});

// 批量事件上报接口
events.post('/batch', zValidator('json', batchEventSchema), async (c) => {
  const { events: rawEvents } = c.req.valid('json');

  // 数据验证和清洗
  const cleanedEvents = analyticsService.cleanEventData(rawEvents as any[]);

  // 转发到Golang服务
  const golangResult = await golangService.postEvents(cleanedEvents);

  // 后台任务：更新实时缓存
  // 注意：在Hono中我们不能直接使用BackgroundTasks，需要用其他方式实现
  setTimeout(() => {
    analyticsService.invalidateDashboardCache();
  }, 0);

  const response: EventResponse = {
    status: 'success',
    message: `成功处理 ${cleanedEvents.length} 个事件`,
    processed_count: cleanedEvents.length,
    golang_response: golangResult,
  };

  return c.json(response);
});

export default events; 